Possibility to set custom calendar options

chdecultot opened the pull request:
Hi,

This PR is a proposal to give users the possibility to set custom options for the calendar views:

1. They can choose to indicate the current time and to show weekends or not:
![image](https://user-images.githubusercontent.com/4903591/43964365-fafd7b52-9cbc-11e8-8a83-6891b346ad61.png)

![image](https://user-images.githubusercontent.com/4903591/43964133-736de00a-9cbc-11e8-9d5e-9df06df9fe73.png)

EDIT: After review, the current time indicator is no longer configurable, but set as default.

2. I have added a 'Today' button between the arrows to allow users to navigate back to the current date quickly (it can be quite annoying without this button if you are far in the future for example).
![image](https://user-images.githubusercontent.com/4903591/43964336-ead6b3c4-9cbc-11e8-8ba4-b06962b4e65c.png)


3. I have also added a recording of the last defaultView the user has clicked on: It is also very annoying today when you create a new event in a Day view that the system shows back the Month view when you are on the calendar view again.

4. Lastly I have changed the translation of the calendar to allow a modification of the words order for languages other than English: e.g in French "Event Calendar" is "Calendrier Evènement"

Please let me know if you are ok with this approach.

Thank you!

chdecultot committed:
Possibility to set custom calendar options

chdecultot committed:
Codacy corrections

netchampfaris commented:
@chdecultot Hey, these are great changes. But, I am not sure if we should store the calendar settings in User doctype. Shouldn't we store those settings in UserSettings ?

chdecultot committed:
Codacy correction

chdecultot commented:
Thanks @netchampfaris,

Do you mean creating a new Doctype User Settings  for these kind of settings ?

In my opinion and from my discussion with users of the calendar in ERPNext and other softwares, calendar settings are user dependent: each person wants his/her own settings depending on their preferences. That's why I registered the settings in the User Doctype.

If we move them to a User Settings Doctype, then I guess these settings will be applicable to all users, isn't it ?

I would prefer a user-wise solution, but let me know your decision.

Thank you!

chdecultot committed:
Codacy corrections

rmehta commented:
@chdecultot lets make current time default... too many settings is not a good solution :)

Why would you not want to see weekends?

chdecultot commented:
Hi @rmehta,

Ok I will make the current time default.
Regarding weekends, the purpose it to reduce the number of columns in order to have a better overview of the calendar in weekly mode.

The issue here is one that has already been discussed many times on the forum: the bootstrap container leads to a very small screen, even on bigger monitor displays. It's not convenient at all to work with, especially when using the calendar.
If any one is not working on weekends, or almost never using the calendar on weekends like me, it is an acceptable solution to disable weekends so you have less of a mess with your events on your screen.

In any case, in terms of UX it's always better to display only what's necessary I guess :smiley:

chdecultot committed:
Set now indicator as default

rmehta commented:
@chdecultot how about a checkbox on the footer of the calendar to show/hide weekends? Contextual settings are much more discoverable!

chdecultot committed:
Weekend options as a button

chdecultot committed:
Codacy corrections

chdecultot commented:
Hi @rmehta,

You are totally right, contextual settings are more discoverable.

I have added a button at the bottom of the calendar, so that each user can set it up directly from the view.
I have also hidden all options from the user doctype, but kept the fields so that these options (weekends + default view) are set on a per user basis.

Instead of a checkbox I have added a button with an explicit label to be in line with Frappe's standard behavior.

Here is the result:
![image](https://user-images.githubusercontent.com/4903591/44092149-e4f3da34-9fcf-11e8-8d44-49594e54b819.png)

![image](https://user-images.githubusercontent.com/4903591/44092173-f1daf1b0-9fcf-11e8-8637-bd3c70eb847f.png)


Let me know if your are OK with these changes.

Thank you!

system19 commented:
@chdecultot I love the Today Button between the arrows, Really useful.
The show/hide weekends button at lower right also looks good. Adding the ability to show / hide outside office hours would be great. Most companies  only need to see 7AM to 7PM. The other 12 hours displayed just makes viewing and moving appointments difficult. Don't know how to connect it through ERPNext but fullcalendar has minTime/maxTime settings to provide this feature. https://fullcalendar.io/docs/minTime

> I have also added a recording of the last defaultView the user has clicked on: It is also very annoying today when you create a new event in a Day view that the system shows back the Month view when you are on the calendar view again.

I used to suffer from this issue but it stopped about a year ago when fullcalendar got updated in a Frappe update.

rmehta commented on frappe/core/doctype/user/user.json:
```diff
@@ -1316,10 +1401,110 @@
+   "collapsible": 0, 
+   "columns": 0, 
+   "default": "month", 
+   "fieldname": "calendar_defaultview", 
```
maybe can store this in `localStorage` too?

chdecultot commented:
I couldn't agree more!

No problem I will move the settings to `localStorage`.
Thanks!

chdecultot committed:
Move preferences to localStorage

chdecultot committed:
Codacy corrections

chdecultot commented:
Hi @rmehta,

I have changed the settings storage to localStorage.
Regarding the proposal of @system19 to implement business hours, what do you think ?

Business hours could be set company wise, but also on a per user basis (everyone may not work during the same hours).
We could also define a standard calendar (everyday 7am to 5pm) or a more granular daily calendar (Monday: 7am-5pm, Tuesday: 8am-6pm, Wednesday: Off, etc...)

For a very granular setup, here is the kind of child table that we could add in the User DocType:
![image](https://user-images.githubusercontent.com/4903591/44146622-c9359e74-a090-11e8-884c-24c48437fde8.png)
Maybe by adding a fallback in the global settings it could be flexible and not too overwhelming in terms of setup ?

Let me know your thoughts and I'll be happy to open another PR on the topic.

Thanks a lot!

system19 commented:
I started to wonder how to stop people from making appointments outside of the minTime/maxTime view settings. FullCalendar has selectConstraint. Would this work in stopping out of view events being created? If not then maybe it could get confusing for user?

rmehta committed:
Update calendar.js

agritheory commented:
@taherkhalil There's a very underused feature in ERPNext called Company Holiday List that follows the same sorts of designs but with dates in mind. It seems to me that business logic regard dates and business logic regarding times should be separated but kept in similar structures.
